File: go/test/planet_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/voxgig-solardemo-sdk/go"
	"github.com/voxgig-sdk/voxgig-solardemo-sdk/go/core"

	vs "github.com/voxgig-sdk/voxgig-solardemo-sdk/go/utility/struct"
)

func TestPlanetEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Planet(nil)
		if ent == nil {
			t.Fatal("expected non-nil PlanetEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"planet": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Planet(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Planet(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := planetBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "planet." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SOLARDEMO_TEST_PLANET_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		planetRef01Ent := client.Planet(nil)
		planetRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "planet"}, setup.data), "planet_ref01"))

		planetRef01DataResult, err := planetRef01Ent.Create(planetRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		planetRef01Data = core.ToMapAny(entityData(planetRef01DataResult))
		if planetRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if planetRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		planetRef01Match := map[string]any{}

		planetRef01ListResult, err := planetRef01Ent.List(planetRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		planetRef01List, planetRef01ListOk := planetRef01ListResult.([]any)
		if !planetRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", planetRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(planetRef01List), map[string]any{"id": planetRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		planetRef01DataUp0Up := map[string]any{
			"id": planetRef01Data["id"],
		}

		planetRef01MarkdefUp0Name := "kind"
		planetRef01MarkdefUp0Value := fmt.Sprintf("Mark01-planet_ref01_%d", setup.now)
		planetRef01DataUp0Up[planetRef01MarkdefUp0Name] = planetRef01MarkdefUp0Value

		planetRef01ResdataUp0Result, err := planetRef01Ent.Update(planetRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		planetRef01ResdataUp0 := core.ToMapAny(entityData(planetRef01ResdataUp0Result))
		if planetRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if planetRef01ResdataUp0["id"] != planetRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if planetRef01ResdataUp0[planetRef01MarkdefUp0Name] != planetRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", planetRef01MarkdefUp0Name, planetRef01ResdataUp0[planetRef01MarkdefUp0Name])
		}

		// LOAD
		planetRef01MatchDt0 := map[string]any{
			"id": planetRef01Data["id"],
		}
		planetRef01DataDt0Loaded, err := planetRef01Ent.Load(planetRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		planetRef01DataDt0LoadResult := core.ToMapAny(entityData(planetRef01DataDt0Loaded))
		if planetRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if planetRef01DataDt0LoadResult["id"] != planetRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		planetRef01MatchRm0 := map[string]any{
			"id": planetRef01Data["id"],
		}
		_, err = planetRef01Ent.Remove(planetRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		planetRef01MatchRt0 := map[string]any{}

		planetRef01ListRt0Result, err := planetRef01Ent.List(planetRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		planetRef01ListRt0, planetRef01ListRt0Ok := planetRef01ListRt0Result.([]any)
		if !planetRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", planetRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(planetRef01ListRt0), map[string]any{"id": planetRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func planetBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "planet", "PlanetTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read planet test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse planet test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"planet01", "planet02", "planet03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SOLARDEMO_TEST_PLANET_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SOLARDEMO_TEST_PLANET_ENTID": idmap,
		"SOLARDEMO_TEST_LIVE":      "FALSE",
		"SOLARDEMO_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["SOLARDEMO_TEST_PLANET_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SOLARDEMO_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewSolardemoSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SOLARDEMO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SOLARDEMO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/planet_entity_test.rb
# Planet entity test

require "minitest/autorun"
require "json"
require_relative "../Solardemo_sdk"
require_relative "runner"

class PlanetEntityTest < Minitest::Test
  def test_create_instance
    testsdk = SolardemoSDK.test(nil, nil)
    ent = testsdk.Planet(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "planet" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = SolardemoSDK.test(seed, nil)
    seen = base.Planet(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = SolardemoConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = SolardemoSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Planet(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = planet_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "planet." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SOLARDEMO_TEST_PLANET_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    planet_ref01_ent = client.Planet(nil)
    planet_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.planet"), "planet_ref01"))

    planet_ref01_data_result = planet_ref01_ent.create(planet_ref01_data, nil)
    planet_ref01_data = Helpers.to_map(planet_ref01_data_result.respond_to?(:data_get) ? planet_ref01_data_result.data_get : planet_ref01_data_result)
    assert !planet_ref01_data.nil?
    assert !planet_ref01_data["id"].nil?

    # LIST
    planet_ref01_match = {}

    planet_ref01_list_result = planet_ref01_ent.list(planet_ref01_match, nil)
    assert planet_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(planet_ref01_list_result),
      { "id" => planet_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    planet_ref01_data_up0_up = {
      "id" => planet_ref01_data["id"],
    }

    planet_ref01_markdef_up0_name = "kind"
    planet_ref01_markdef_up0_value = "Mark01-planet_ref01_#{setup[:now]}"
    planet_ref01_data_up0_up[planet_ref01_markdef_up0_name] = planet_ref01_markdef_up0_value

    planet_ref01_resdata_up0_result = planet_ref01_ent.update(planet_ref01_data_up0_up, nil)
    planet_ref01_resdata_up0 = Helpers.to_map(planet_ref01_resdata_up0_result.respond_to?(:data_get) ? planet_ref01_resdata_up0_result.data_get : planet_ref01_resdata_up0_result)
    assert !planet_ref01_resdata_up0.nil?
    assert_equal planet_ref01_resdata_up0["id"], planet_ref01_data_up0_up["id"]
    assert_equal planet_ref01_resdata_up0[planet_ref01_markdef_up0_name], planet_ref01_markdef_up0_value

    # LOAD
    planet_ref01_match_dt0 = {
      "id" => planet_ref01_data["id"],
    }
    planet_ref01_data_dt0_loaded = planet_ref01_ent.load(planet_ref01_match_dt0, nil)
    planet_ref01_data_dt0_load_result = Helpers.to_map(planet_ref01_data_dt0_loaded.respond_to?(:data_get) ? planet_ref01_data_dt0_loaded.data_get : planet_ref01_data_dt0_loaded)
    assert !planet_ref01_data_dt0_load_result.nil?
    assert_equal planet_ref01_data_dt0_load_result["id"], planet_ref01_data["id"]

    # REMOVE
    planet_ref01_match_rm0 = {
      "id" => planet_ref01_data["id"],
    }
    planet_ref01_ent.remove(planet_ref01_match_rm0, nil)

    # LIST
    planet_ref01_match_rt0 = {}

    planet_ref01_list_rt0_result = planet_ref01_ent.list(planet_ref01_match_rt0, nil)
    assert planet_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(planet_ref01_list_rt0_result),
      { "id" => planet_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def planet_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "planet", "PlanetTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = SolardemoSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["planet01", "planet02", "planet03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SOLARDEMO_TEST_PLANET_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SOLARDEMO_TEST_PLANET_ENTID" => idmap,
    "SOLARDEMO_TEST_LIVE" => "FALSE",
    "SOLARDEMO_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["SOLARDEMO_TEST_PLANET_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SOLARDEMO_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = SolardemoSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SOLARDEMO_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SOLARDEMO_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
